File: 2015/rust/src/day12.rs
use serde_json::Value;
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;

pub fn day12(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 12: JSAbacusFramework.io");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    println!("\tPart 1: {}", part1(&input)?);
    println!("\tPart 2: {}", part2(&input)?);
    Ok(())
}

fn json_sum(json: &Value) -> i32 {
    match json {
        Value::Null | Value::Bool(_) | Value::String(_) => 0,
        Value::Number(num) => num.as_i64().unwrap_or(0) as i32,
        Value::Object(obj) => obj.values().fold(0, |sum, x| sum + json_sum(x)),
        Value::Array(arr) => arr.iter().fold(0, |sum, x| sum + json_sum(x)),
    }
}

fn part1(input: &str) -> Result<i32, Box<dyn Error>> {
    Ok(json_sum(&serde_json::from_str(input)?))
}

fn json_sum_no_red(json: &Value) -> i32 {
    match json {
        Value::Null | Value::Bool(_) | Value::String(_) => 0,
        Value::Number(num) => num.as_i64().unwrap_or(0) as i32,
        Value::Object(obj) => match obj.values().any(|x| match x {
            Value::String(s) if s == "red" => true,
            _ => false,
        }) {
            true => 0,
            false => obj.values().fold(0, |sum, x| sum + json_sum_no_red(x)),
        },
        Value::Array(arr) => arr.iter().fold(0, |sum, x| sum + json_sum_no_red(x)),
    }
}

fn part2(input: &str) -> Result<i32, Box<dyn Error>> {
    Ok(json_sum_no_red(&serde_json::from_str(input)?))
}

File: 2015/go/day12/day12.go
package day12

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
)

//Day12 JSAbacusFramework.io
func Day12(inputFile string) error {
	input, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return err
	}

	fmt.Println("Day 12: JSAbacusFramework.io")
	part1, err := part1(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 1: %d\n", part1)

	part2, err := part2(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 2: %d\n", part2)

	return nil
}

func sumJSON(j interface{}, filter string) int {
	sum := 0
	switch v := j.(type) {
	case float64:
		sum += int(v)
	case int:
		sum += v
	case []interface{}:
		for _, vv := range v {
			sum += sumJSON(vv, filter)
		}
	case map[string]interface{}:
		if filter != "" {
			for _, vv := range v {
				if vvv, ok := vv.(string); ok {
					if vvv == filter {
						return 0
					}
				}
			}
		}
		for _, vv := range v {
			sum += sumJSON(vv, filter)
		}
	}
	return sum
}

func part1(input string) (int, error) {
	var js interface{}
	if err := json.Unmarshal([]byte(input), &js); err != nil {
		return 0, err
	}

	return sumJSON(js, ""), nil
}

func part2(input string) (int, error) {
	var js interface{}
	if err := json.Unmarshal([]byte(input), &js); err != nil {
		return 0, err
	}

	return sumJSON(js, "red"), nil
}
